csaf-import/src/main/kotlin/io/github/csaf/sbom/retrieval/RetrievedProvider.kt

```diff
@@ -57,44 +56,42 @@ class RetrievedProvider(val json: Provider) {
     fun validate(validationContext: ValidationContext) {
         role.checkRole(validationContext).let { vr ->
             if (vr is ValidationFailed) {
-                throw ValidationException(vr)
+                throw vr.toException()
             }
         }
     }
 
     /** This function fetches all CSAF documents that are listed by this provider. */
     suspend fun fetchDocuments(loader: CsafLoader = lazyLoader): List<Result<RetrievedDocument>> {
         @Suppress("SimpleRedundantLet")
-        return json.distributions?.let { distributions ->
-            distributions
-                .mapNotNull { distribution ->
-                    distribution.directory_url?.let { it.toString().trimEnd('/') }
-                }
-                .mapAsync { directoryUrl ->
-                    val indexUrl = "$directoryUrl/index.txt"
-                    loader
-                        .fetchText(indexUrl)
-                        .fold(
-                            { index ->
-                                index.lines().mapAsync { line ->
-                                    val csafUrl = "$directoryUrl/$line"
-                                    RetrievedDocument.from(csafUrl, loader, this.role)
-                                }
-                            },
-                            { e ->
-                                listOf(
-                                    Result.failure(
-                                        Exception(
-                                            "Failed to fetch index.txt from directory at $directoryUrl",
-                                            e
-                                        )
+        return (json.distributions ?: emptySet())
```

Comment:
Did the change make it actually concurrent now? It seems that before that it actually only returns when all documents are fetched instead of yielding the already finished documents. Is `flatten()` correct here?

Reply:
The process was highly concurrent before and it still is. This applies to both the directory level and the document level. Semantics are unchanged. `flatten()` is used to combine the results of all directory URLs (one `List<Result<RetrievedDocument>>` per directory) into one concatenated `List<Result<RetrievedDocument>>`.

Reply:
If I run the main application with a large provider, e.g. www.redhat.com then you can see that the `onSuccess` of the provider returns immediate but the `onSuccess` of the individual documents is only called once ALL documents are retrieved, which is not what I would expect. What's the idiomatic way to return a list in a suspend function? It seems that we basically turn the suspend function into a blocking call from the caller perspective with the `flatten`.

Reply:
See https://github.com/csaf-sbom/kotlin-csaf/issues/59